### PR TITLE
add implementation for dsv4-mhc

### PR DESCRIPTION
**Pull request by xiaolong-intel**
# Essential Elements of an Effective PR Description Checklist

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS ABOVE HAVE BEEN CONSIDERED.

## Purpose
DeepSeek-V4 introduces the manifold-constrained Hyper-Connection (mHC) mechanism to improve feature flow across layers. The current vLLM XPU backend lacks native implementations for the mHC pre/post processing kernels, causing the model to fall back to slow PyTorch composite operations on Intel GPUs.
This PR adds high-performance SYCL-based device kernels for the three core mHC operators—mhc_pre, mhc_post, and hc_head_fused—optimized specifically for Intel Arc Pro B60 and compatible with the existing vLLM XPU operator registration framework.
## Test Plan
Use tests/test_mhc.py to test the accuracy of the operator under bf16 input
Use benchmark/benchmark_mhc.py to test the operator's running speed
## Test Result
accurency：
root@B60-5468-1:/xiaolong/vllm-xpu-kernels-internal# PYTHONPATH=./ pytest -q tests/test_mhc.py
...................................... [100%]
38 passed in 2.43s
speedup：
image
## (Optional) Documentation Update

**BEFORE SUBMITTING, PLEASE READ <https://docs.vllm.ai/en/latest/contributing>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: csrc/xpu/mhc/mhc_pre.cpp [hunk 1]
```diff
@@ -0,0 +1,791 @@
...
+                    for (int off = HC; off < COMB_LANES; off <<= 1)
+                      cs += sycl::permute_group_by_xor(sg, cs, off);
+                    v *= sycl::native::recip(cs + sk_eps);
+
+                    if (it_sk < sk_repeat - 1) {
+                      float rs = v;
+#pragma unroll
+                      for (int off = 1; off < HC; off <<= 1)
+                        rs += sycl::permute_group_by_xor(sg, rs, off);
+                      v *= sycl::native::recip(rs + sk_eps);
```
Comment (xinyu-intel) [inline csrc/xpu/mhc/mhc_pre.cpp]: random place, some using sycl::native, some not. Is it expected?

File: csrc/xpu/sycl/deepseek_mhc.cpp [hunk 2]
```diff
@@ -0,0 +1,791 @@
...
+      hc_scale.data_ptr<float>(),
+      hc_base.data_ptr<float>(),
+      reinterpret_cast<bf16*>(out.data_ptr()),
+      static_cast<int>(N),
+      static_cast<int>(hidden_size),
+      static_cast<float>(rms_eps),
+      static_cast<float>(hc_eps));
+}
+
+#undef DISPATCH_HC
```
Comment (xinyu-intel) [inline csrc/xpu/sycl/deepseek_mhc.cpp]: add blank line

File: benchmark/benchmark_mhc.py [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+    for _ in range(warmup_iters):
+        func_callable(*args)
+    torch.xpu.synchronize()
+
+    # Reset memory stats before benchmark
+    reset_memory_stats()
+
+    # Benchmark
+    start_events = [torch.xpu.Event(enable_timing=True) for _ in range(benchmark_iters)]
+    end_events = [torch.xpu.Event(enable_timing=True) for _ in range(benchmark_iters)]
```
Comment (xinyu-intel) [inline benchmark/benchmark_mhc.py]: please use single event for start and end instead.

File: benchmark/benchmark_mhc.py [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+
+    for i in range(benchmark_iters):
+        start_events[i].record()
+        func_callable(*args)
+        end_events[i].record()
```
Comment (xinyu-intel) [inline benchmark/benchmark_mhc.py]:
```suggestion
    start_events.record()
    for i in range(benchmark_iters):
        func_callable(*args)
    end_events[i].record()
```

File: csrc/xpu/mhc/mhc_pre.cpp (hunk 1 above)
Comment (xiaolong-intel) [reply csrc/xpu/mhc/mhc_pre.cpp]: I think this is intentional on my part. My understanding is that sycl::native::* is only available for a specific subset of math functions (like exp, rsqrt, recip), and I used it only in the activation/normalization paths where the performance gain seems to matter most and the accuracy drop is likely acceptable for inference.
For the other operations:
sycl::fma appears to be a hardware-native fused instruction already, so I'm not sure a native variant even exists.
Basic arithmetic and group primitives (reduce, permute, max) seem to be directly mapped by the compiler and, as far as I know, don't have approximate "native" versions.
That said, I haven't verified every SYCL built-in table exhaustively, so if you know of a native variant I missed, or if there's a project policy against mixing them, I'm happy to align everything to the standard precision versions or document the rationale more explicitly.

File: benchmark/benchmark_mhc.py (hunk 3 above)
Comment (xiaolong-intel) [reply benchmark/benchmark_mhc.py]: Okay, I will go and correct it now, and I will pay attention next time I use it. Thank you for your suggestion.

File: csrc/xpu/ops.h [hunk 5]
```diff
@@ -85,6 +85,34 @@ void multimodal_rotary_embedding(
...
+    const at::Tensor& hs_flat,
+    const at::Tensor& fn,
+    const at::Tensor& hc_scale,
+    const at::Tensor& hc_base,
+    at::Tensor& out,
+    int64_t hidden_size,
+    double rms_eps,
+    double hc_eps,
+    int64_t hc_mult);
+
```
Comment (majian4work) [inline csrc/xpu/ops.h:115]: Could you also add this new fused kernel? https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/mhc.py#L306

File: csrc/xpu/sycl/deepseek_mhc.cpp [hunk 6]
```diff
@@ -0,0 +1,791 @@
...
+          }
+        });
+  });
+}
+
+#define DISPATCH_HC(HC_VAR, NAME, ...)       \
+  do {                                       \
+    switch (HC_VAR) {                        \
+      case 2:                                \
+        NAME<2>(__VA_ARGS__);                \
```
Comment (YangQun1) [inline csrc/xpu/sycl/deepseek_mhc.cpp]: why do we need to support HC=2? dsv4 seems to only use HC=4

**Comment by jikunshang** [comment]
do you ever build on your side? 
error in ci:
```
2026-05-29T02:26:40.6503585Z ccache /opt/intel/oneapi/compiler/2025.3/bin/icpx -DPy_LIMITED_API=3 -DTORCH_EXTENSION_NAME=_xpu_C -DUSE_C10D_GLOO -DUSE_C10D_XCCL -DUSE_DISTRIBUTED -DUSE_RPC -DUSE_TENSORPIPE -D_xpu_C_EXPORTS -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/third_party/oneDNN/include -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/build/temp/oneDNN/include -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/third_party/oneDNN/src -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/third_party/oneDNN/src/../include -I/opt/intel/oneapi/compiler/2025.3/include/syclcompat -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/grouped_gemm/xe_default -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/grouped_gemm/xe_default/.. -I/__w/vllm-xpu-kernels/vllm-xpu-kernels -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/grouped_gemm/xe_2 -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/grouped_gemm/xe_2/.. -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/gdn_attn/xe_2 -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/gdn_attn/xe_2/.. -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/mqa_logits/xe_2 -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/mqa_logits/xe_2/.. -isystem /opt/intel/oneapi/compiler/2025.3/include -isystem /opt/intel/oneapi/compiler/2025.3/include/sycl -isystem /usr/include/python3.12 -isystem /opt/venv/lib/python3.12/site-packages/torch/include -isystem /opt/venv/lib/python3.12/site-packages/torch/include/torch/csrc/api/include -Werror -include /__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/sycl_first.h -O3 -DNDEBUG -std=gnu++20 -fPIC -fsycl -O3 -DNDEBUG -fsycl-targets=spir64_gen -DVLLM_XPU_ENABLE_XE_DEFAULT -DVLLM_XPU_ENABLE_XE2 -DVLLM_MOE_ENABLED -DVLLM_GDN_ENABLED -DVLLM_MQA_LOGITS_ENABLED -MD -MT CMakeFiles/_xpu_C.dir/csrc/xpu/sycl/deepseek_mhc.cpp.o -MF CMakeFiles/_xpu_C.dir/csrc/xpu/sycl/deepseek_mhc.cpp.o.d -o CMakeFiles/_xpu_C.dir/csrc/xpu/sycl/deepseek_mhc.cpp.o -c /__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/sycl/deepseek_mhc.cpp
2026-05-29T02:26:40.6527439Z ../../csrc/xpu/sycl/deepseek_mhc.cpp:621:59: error: no member named 'xpu' in namespace 'vllm'
2026-05-29T02:26:40.6542837Z   621 | static inline sycl::queue& current_queue() { return vllm::xpu::vllmGetQueue(); }
2026-05-29T02:26:40.6544599Z       |                                                     ~~~~~~^
2026-05-29T02:26:40.6545838Z 1 error generated.
```

File: csrc/xpu/ops.h (hunk 5 above)
Comment (YangQun1) [reply csrc/xpu/ops.h:115]: I had a fused kernel impl and some optimize for mhc_pre. After this PR merged, I will rebase and submit another PR to address them. aligned with @xiaolong-intel

File: csrc/xpu/sycl/deepseek_mhc.cpp (hunk 6 above)
Comment (xiaolong-intel) [reply csrc/xpu/sycl/deepseek_mhc.cpp]: That HC=2 was added by mistake; indeed, there are only dsv4 conditions with HC=4. I will delete it later.Currently, deepseek-mhc.cpp has been split into three files under the csrc/xpu/mhc path according to functionality.

**Comment by xiaolong-intel** [comment]
> do you ever build on your side? error in ci:
> 
> ```
> 2026-05-29T02:26:40.6503585Z ccache /opt/intel/oneapi/compiler/2025.3/bin/icpx -DPy_LIMITED_API=3 -DTORCH_EXTENSION_NAME=_xpu_C -DUSE_C10D_GLOO -DUSE_C10D_XCCL -DUSE_DISTRIBUTED -DUSE_RPC -DUSE_TENSORPIPE -D_xpu_C_EXPORTS -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/third_party/oneDNN/include -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/build/temp/oneDNN/include -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/third_party/oneDNN/src -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/third_party/oneDNN/src/../include -I/opt/intel/oneapi/compiler/2025.3/include/syclcompat -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/grouped_gemm/xe_default -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/grouped_gemm/xe_default/.. -I/__w/vllm-xpu-kernels/vllm-xpu-kernels -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/grouped_gemm/xe_2 -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/grouped_gemm/xe_2/.. -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/gdn_attn/xe_2 -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/gdn_attn/xe_2/.. -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/mqa_logits/xe_2 -I/__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/mqa_logits/xe_2/.. -isystem /opt/intel/oneapi/compiler/2025.3/include -isystem /opt/intel/oneapi/compiler/2025.3/include/sycl -isystem /usr/include/python3.12 -isystem /opt/venv/lib/python3.12/site-packages/torch/include -isystem /opt/venv/lib/python3.12/site-packages/torch/include/torch/csrc/api/include -Werror -include /__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/sycl_first.h -O3 -DNDEBUG -std=gnu++20 -fPIC -fsycl -O3 -DNDEBUG -fsycl-targets=spir64_gen -DVLLM_XPU_ENABLE_XE_DEFAULT -DVLLM_XPU_ENABLE_XE2 -DVLLM_MOE_ENABLED -DVLLM_GDN_ENABLED -DVLLM_MQA_LOGITS_ENABLED -MD -MT CMakeFiles/_xpu_C.dir/csrc/xpu/sycl/deepseek_mhc.cpp.o -MF CMakeFiles/_xpu_C.dir/csrc/xpu/sycl/deepseek_mhc.cpp.o.d -o CMakeFiles/_xpu_C.dir/csrc/xpu/sycl/deepseek_mhc.cpp.o -c /__w/vllm-xpu-kernels/vllm-xpu-kernels/csrc/xpu/sycl/deepseek_mhc.cpp
> 2026-05-29T02:26:40.6527439Z ../../csrc/xpu/sycl/deepseek_mhc.cpp:621:59: error: no member named 'xpu' in namespace 'vllm'
> 2026-05-29T02:26:40.6542837Z   621 | static inline sycl::queue& current_queue() { return vllm::xpu::vllmGetQueue(); }
> 2026-05-29T02:26:40.6544599Z       |                                                     ~~~~~~^
> 2026-05-29T02:26:40.6545838Z 1 error generated.
> ```

Sorry, utils.h was missing before. I re-downloaded the code and compiled it, and now the compilation has no problem.

**Review by YangQun1 — APPROVED** [review]

File: csrc/xpu/mhc/hc_head_fused.cpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+  sycl::group_barrier(it.get_group());
+
+#pragma unroll
+  for (int j = 0; j < COUNT; ++j)
+    results[j] = scratch[j];
+}
+
+class HcHeadKernel;
+
+static sycl::event launch_hc_head_fused(
```
Comment (jikunshang) [inline csrc/xpu/mhc/hc_head_fused.cpp]: please use functor style like https://github.com/vllm-project/vllm-xpu-kernels/pull/18

File: csrc/xpu/mhc/mhc_post.cpp [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+  static constexpr int VEC_POST = 8;
+  static constexpr int WG_SIZE = 256;
+
+  sycl::range<1> global(static_cast<size_t>(N) * WG_SIZE);
+  sycl::range<1> local(WG_SIZE);
+
+  return q.submit([&](sycl::handler& h) {
+    sycl::local_accessor<float, 1> coeff_s(HC * HC + HC, h);
+
+    h.parallel_for<MhcPostKernel>(
```
Comment (jikunshang) [inline csrc/xpu/mhc/mhc_post.cpp]: i think we can just ignore `<MhcPostKernel>`

File: csrc/xpu/mhc/mhc_pre.cpp [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+#include "utils.h"
+
+namespace vllm {
+
+using bf16 = sycl::ext::oneapi::bfloat16;
+
+static constexpr int HC = 4;
+
+template <int N>
+static inline sycl::vec<bf16, N> vload_bf16(const bf16* p) {
```
Comment (jikunshang) [inline csrc/xpu/mhc/mhc_pre.cpp]: it's duplicated in mhc_post.cpp,  we can move to utils.h

File: csrc/xpu/mhc/mhc_pre.cpp [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+                      for (int off = 1; off < HC; off <<= 1)
+                        rs += sycl::permute_group_by_xor(sg, rs, off);
+                      v *= sycl::native::recip(rs + sk_eps);
+                    }
+                  }
+                  comb_mix[tok * HC * HC + my_idx] = v;
+                }
+              }
+
+              if (round < NUM_COMB_ROUNDS - 1) sycl::group_barrier(sg);
```
Comment (jikunshang) [inline csrc/xpu/mhc/mhc_pre.cpp]:
```suggestion
              if (round < NUM_COMB_ROUNDS - 1) {
                  sycl::group_barrier(sg);
              }
```

File: csrc/xpu/mhc/mhc_pre.cpp [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+                    float cs = v;
+#pragma unroll
+                    for (int off = HC; off < COMB_LANES; off <<= 1)
+                      cs += sycl::permute_group_by_xor(sg, cs, off);
+                    v *= sycl::native::recip(cs + sk_eps);
+
+                    if (it_sk < sk_repeat - 1) {
+                      float rs = v;
+#pragma unroll
+                      for (int off = 1; off < HC; off <<= 1)
```
Comment (jikunshang) [inline csrc/xpu/mhc/mhc_pre.cpp]: it's a 5 level nested loop? does it make sense? it's less readable.
